File: c_code/lexer.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define MAX_TOKEN_SIZE 100

// Token types
typedef enum
{
    TOKEN_IDENTIFIER,
    TOKEN_NUMBER,
    TOKEN_STRING,
    TOKEN_KEYWORD,
    TOKEN_OPERATOR,
    TOKEN_PUNCTUATION,
    TOKEN_NEWLINE,
    TOKEN_EOF
} TokenType;

// Token structure
typedef struct
{
    TokenType type;
    char value[MAX_TOKEN_SIZE];
} Token;

// Function to check if a string is a keyword
int is_keyword(const char *str)
{
    const char *keywords[] = {"let", "print", "for", "if", "else"};
    int num_keywords = sizeof(keywords) / sizeof(keywords[0]);

    for (int i = 0; i < num_keywords; i++)
    {
        if (strcmp(keywords[i], str) == 0)
        {
            return 1;
        }
    }
    return 0;
}

// Lexer function
Token *lexer(const char *input)
{
    Token *tokens = malloc(sizeof(Token) * MAX_TOKEN_SIZE);
    int token_count = 0;

    const char *ptr = input;
    while (*ptr)
    {
        // Skip whitespace
        if (isspace(*ptr))
        {
            ptr++;
            continue;
        }

        // Recognize numbers
        if (isdigit(*ptr))
        {
            tokens[token_count].type = TOKEN_NUMBER;
            int i = 0;
            while (isdigit(ptr[i]))
            {
                tokens[token_count].value[i] = ptr[i];
                i++;
            }
            tokens[token_count].value[i] = '\0';
            token_count++;
            ptr += i;
        }
        // Recognize keywords
        else if (isalpha(*ptr))
        {
            int i = 0;
            while (isalnum(ptr[i]))
            {
                tokens[token_count].value[i] = ptr[i];
                i++;
            }
            tokens[token_count].value[i] = '\0';

            if (is_keyword(tokens[token_count].value))
            {
                tokens[token_count].type = TOKEN_KEYWORD;
            }
            else
            {
                tokens[token_count].type = TOKEN_IDENTIFIER;
            }

            token_count++;
            ptr += i;
        }
        // Handle other cases (operators, punctuation, etc.)
        else
        {
            // You would add more code to recognize operators and other constructs here.
            // For this simple example, we'll just skip the character.
            ptr++;
        }
    }

    // Add an end-of-file token
    tokens[token_count].type = TOKEN_EOF;

    return tokens;
}

int main()
{
    const char *input = "let x = 42; print(x);";
    Token *tokens = lexer(input);

    int i = 0;
    while (tokens[i].type != TOKEN_EOF)
    {
        printf("Token: %s\n", tokens[i].value);
        i++;
    }

    free(tokens);
    return 0;
}

File: js_code/lexer.js
// Token types
const TokenType = {
    IDENTIFIER: 'IDENTIFIER',
    NUMBER: 'NUMBER',
    STRING: 'STRING',
    KEYWORD: 'KEYWORD',
    OPERATOR: 'OPERATOR',
    PUNCTUATION: 'PUNCTUATION',
    NEWLINE: 'NEWLINE',
    EOF: 'EOF',
};

// Token structure
class Token {
    constructor(type, value) {
        this.type = type;
        this.value = value;
    }
}

// Function to check if a string is a keyword
function isKeyword(str) {
    const keywords = ['let', 'print', 'for', 'if', 'else'];
    return keywords.includes(str);
}

// Lexer function
function lexer(input) {
    const tokens = [];
    let i = 0;

    while (i < input.length) {
        // Skip whitespace
        if (/\s/.test(input[i])) {
            i++;
            continue;
        }

        // Recognize numbers
        if (/[0-9]/.test(input[i])) {
            let j = i;
            while (/[0-9]/.test(input[j])) {
                j++;
            }
            tokens.push(new Token(TokenType.NUMBER, input.slice(i, j)));
            i = j;
        }

        // Recognize keywords and identifiers
        if (/[a-zA-Z_]/.test(input[i])) {
            let j = i;
            while (/[a-zA-Z0-9_]/.test(input[j])) {
                j++;
            }
            const tokenValue = input.slice(i, j);
            tokens.push(new Token(isKeyword(tokenValue) ? TokenType.KEYWORD : TokenType.IDENTIFIER, tokenValue));
            i = j;
        }

        // Recognize operators
        if (/[\+\-\*\/\=\&\|\<\>\!\%\^]/.test(input[i])) {
            tokens.push(new Token(TokenType.OPERATOR, input[i]));
            i++;
        }

        // Recognize punctuation
        if (/[\(\)\{\}\[\]\,\;\:\.]/.test(input[i])) {
            tokens.push(new Token(TokenType.PUNCTUATION, input[i]));
            i++;
        }

        // Recognize newline characters
        if (/\n/.test(input[i])) {
            tokens.push(new Token(TokenType.NEWLINE, '\n'));
            i++;
        }
    }

    // Add an end-of-file token
    tokens.push(new Token(TokenType.EOF, ''));

    return tokens;
}

module.exports = { TokenType, Token, lexer };

// Example input
const input = 'let x = 42; print(x);';

const tokens = lexer(input);
console.log(tokens);
